dulwich/porcelain.py

```diff
@@ -2100,113 +2101,167 @@ def checkout_branch(repo, target: Union[bytes, str], force: bool = False) -> Non
+    Writes an initial ``.git/info/sparse-checkout`` file that includes only
+    top-level files (and excludes all subdirectories), e.g. ``["/*", "!/*/"]``.
+    Then performs a sparse checkout to update the working tree accordingly.
+
+    If no directories are specified, then only top-level files are included:
+    https://git-scm.com/docs/git-sparse-checkout#_internalscone_mode_handling
+
+    Args:
+      repo: Path to the repository or a Repo object.
+
+    Returns:
+      None
+    """
+    if not isinstance(repo, Repo):
+        repo = Repo(repo)
+
+    config = repo.get_config()
+    config.set((b"core",), b"sparseCheckout", b"true")
+    config.set((b"core",), b"sparseCheckoutCone", b"true")
+    config.write_to_path()
+
+    patterns = ["/*", "!/*/"]  # root-level files only
+    sparse_checkout(repo, patterns, force=True, cone=True)
+
+
+def cone_mode_set(repo, dirs, force=False):
+    """Overwrite the existing 'cone-mode' sparse patterns with a new set of directories.
+
+    Ensures ``core.sparseCheckout`` and ``core.sparseCheckoutCone`` are enabled.
+    Writes new patterns so that only the specified directories (and top-level files)
+    remain in the working tree, and applies the sparse checkout update.
+
+    Args:
+      repo: Path to the repository or a Repo object.
+      dirs: List of directory names to include.
+      force: Whether to forcibly discard local modifications (default False).
+
+    Returns:
+      None
+    """
+    if not isinstance(repo, Repo):
+        repo = Repo(repo)
+
+    config = repo.get_config()
+    config.set((b"core",), b"sparseCheckout", b"true")
+    config.set((b"core",), b"sparseCheckoutCone", b"true")
+    config.write_to_path()
+
+    # Initial lines: include top-level files, then exclude all subdirectories
+    new_patterns = ["/*", "!/*/"]
+
+    # For each directory to include, add an exclusion-style line that "undoes"
+    # the prior 'exclude' and re-includes that directory (and everything under it).
+    for d in dirs:
+        d = d.strip("/")
+        if d:
+            new_patterns.append(f"!/{d}/")
+
+    # Finally, apply the patterns and update the working tree
+    sparse_checkout(repo, new_patterns, force=force, cone=True)
+
+
+def cone_mode_add(repo, dirs, force=False):
+    """Add new directories to the existing 'cone-mode' sparse-checkout patterns.
+
+    Reads the current patterns from ``.git/info/sparse-checkout``, adds pattern
+    lines to include the specified directories, and then performs a sparse
+    checkout to update the working tree accordingly.
+
+    Args:
+      repo: Path to the repository or a Repo object.
+      dirs: List of directory names to add to the sparse-checkout.
+      force: Whether to forcibly discard local modifications (default False).
+
+    Returns:
+      None
+    """
+    if not isinstance(repo, Repo):
+        repo = Repo(repo)
+
+    sp_path = os.path.join(repo.path, ".git", "info", "sparse-checkout")
```

Comment:
Please move the implementation-specific details to Repo

Reply:
yep, there was actually already a method for accessing this path, reused that existing one